### DISCUSSION TITLE
mx.scan?

### DISCUSSION THREAD

**Discussion by pharringtonp19** [ml-explore/mlx#857] [Q&A] [answered]
Is there an equivalent of `jax.lax.scan` in mx?

**Comment by awni** [comment, marked as answer]
No we don't have that.

**Reply by awni** [reply]
Just curious, what do you want it for?

**Reply by pharringtonp19** [reply]
@awni  Thanks for the response.

I use it as a "compiled for loop" in my training functions (see below). 

```
@dataclass
class Trainer:
    loss_fn: Callable
    opt: optax.GradientTransformation
    epochs: int

    # Train Function
    def train(self, params: Model_Params, X, Y, mask):
        def update_fn(carry, t):
            params, opt_state = carry
            loss_values, grads = jax.value_and_grad(self.loss_fn, has_aux=self.loss_fn.aux_status)(params, X, Y, mask)
            updates, opt_state = self.opt.update(grads, opt_state, params)
            params = optax.apply_updates(params, updates)
            return (params, opt_state), loss_values
        (opt_params, _), loss_values_history = jax.lax.scan(update_fn, (params, self.opt.init(params)), xs=None, length=self.epochs)
        return opt_params, loss_values_history
```

**Reply by awni** [reply]
Makes sense.

Why not compile the `update_fn` directly with `jax.jit` and then put it in a regular Python loop? Is it just a syntax thing or do you notice a speed improvement?

**Reply by pharringtonp19** [reply]
@awni tbh I had to go back and look at the Jax documentation to figure out why I write it with a `scan`. 

One thing that jumps out is that in Jax we cannot reverse mode differentiate through a `for` loop while while we can through a `scan` ([Reference](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.fori_loop.html)). This is important in my context where I am fitting models via bilevel gradient descent and need to "backprop" through the inner for loop.

**Reply by angeloskath** [reply]
@pharringtonp19 in Jax you can backprop through the loop (it is just unrolled). The reason people often use this pattern is because it reduces compilation time significantly, especially if the loop is big. Often compilation even fails due to memory or other issues if the graph is too big.

**Reply by angeloskath** [reply]
On the other hand reading the docs you referenced it seems you are right so ignore the possible noise. I just have seen the scan pattern to reduce compilation time once too many times and thought it applies.

**Reply by awni** [reply]
🤔 I'm pretty sure you can backprop through a loop in jax:

```python
import jax
import jax.numpy as jnp


def loop_fn(x):
    for _ in range(5):
        x = jnp.sin(x)
    return x

print(jax.grad(loop_fn)(jnp.array(1.0)))
```

**Reply by awni** [reply]
If you use a fixed iteration count in the inner loop, I think it's fine. And indeed the only reason I can see to use a scan would be to minimize compile time like @angeloskath said..

**Reply by pharringtonp19** [reply]
@angeloskath @awni That was my mistake. I should just have presented a mwe instead of incorrectly referencing the jax documents. Thanks for your time!
